feature/match-result

Mostafa-khaled21 opened the pull request:
https://github.com/user-attachments/assets/3b13835d-9e26-4172-84d6-80999d48eb72

Mostafa-khaled21 committed:
add matching movies and series use cases and their unit tests

Mostafa-khaled21 committed:
add match remote data source impl and interface

Mostafa-khaled21 committed:
add match repository and its implementation

Mostafa-khaled21 committed:
add match api service retrofit

Mostafa-khaled21 committed:
add login and collection bottom sheets and components

Mostafa-khaled21 committed:
Add new string resources and dependencies

- Add various error messages and UI text strings in English and Arabic.
- Include domain:user and api:authentication project dependencies.

Mostafa-khaled21 committed:
refactor match result screen and add utils

Mostafa-khaled21 committed:
Refactor MatchResult navigation and add LoginRoute

This commit refactors the navigation to the MatchResult screen by removing arguments from its route.
It also introduces a new LoginRoute to handle navigation to the login screen, integrating AuthenticationApi.

Mostafa-khaled21 committed:
add match api service and match response

Mostafa-khaled21 committed:
handle exceptions and enhance match result screen

Mostafa-khaled21 committed:
enhance base view model and add match result screen

Mostafa-khaled21 committed:
refactor MatchPagerScreen and save state in MatchPagerViewModel

Mostafa-khaled21 committed:
pass matching selections to result screen

mohamedebrahem13 commented on presentation/match/src/main/java/com/giraffe/match/screen/result/MatchResultViewModel.kt:
```diff
@@ -0,0 +1,377 @@
+    private var currentSelectedMediaType: MediaType? = null
+
+    init {
+        _genreIds = savedStateHandle.get<List<Int>>("genreSelections")
```
If you used the save state handle approach then you can remove the parameter from on finsh callback

Mostafa-khaled21 replied:
When I deleted them from the route, nothing arrived.

HendSayed25 commented on presentation/match/src/main/res/values-ar-rEG/strings.xml:
```diff
@@ -48,4 +48,27 @@
+    <string name="username_error">لقد حدث شئ ما اثناء جلب اسم المستخدم و حاول مرة اخرى</string>
+    <string name="error_infrastructure">.حدث خطأ داخلي في النظام. يُرجى المحاولة مرة أخرى لاحقًا</string>
+    <string name="you_re_almost_there">أنت على وشك الانتهاء!</string>
+    <string name="log_in_to_save_movies_create_collections_and_get_personalized_recommendations"> </string>
```
where is the text here?

Mostafa-khaled21 replied:
<img width="1692" height="42" alt="image" src="https://github.com/user-attachments/assets/8d263cf8-fe12-4c14-9a2a-388622768996" />

<img width="1692" height="42" alt="image" src="https://github.com/user-attachments/assets/54df0ab8-560b-41af-92b2-1453e9fbb133" />

dev-bilal-azzam commented on datasource/remote/media/src/main/java/com/giraffe/media/match/retrofit/MatchRemoteDataSourceImpl.kt:
```diff
@@ -0,0 +1,44 @@
+import javax.inject.Inject
+
+class MatchRemoteDataSourceImplRetrofit @Inject constructor(
+    private val retrofitRequestBuilder: RetrofitRequestBuilder<MatchApiService>
```
No Need for this RequestzBuilder, just safeCall is enough

HendSayed25 commented on presentation/match/src/main/java/com/giraffe/match/screen/result/MatchResultViewModel.kt:
```diff
@@ -0,0 +1,377 @@
+        updateState { it.copy(isLoading = true, isError = false) }
+
+        safeExecute(
+            onSuccess = { data ->
```
separate the success state to different function for more readability and for all success body if has more than 2 lines

dev-bilal-azzam replied:
consider separating the onSuccess to a separated function.

dev-bilal-azzam commented on domain/media/src/main/java/com/giraffe/media/match/usecase/GetMatchingMoviesUseCase.kt:
```diff
@@ -0,0 +1,25 @@
+import javax.inject.Inject
+
+class GetMatchingMoviesUseCase @Inject constructor(
+    private val repository: MatchRepository
```
I think full name is better ` private val matchRepository: MatchRepository`

dev-bilal-azzam replied:
This comment is not just for this usecase

dev-bilal-azzam commented on presentation/match/src/main/java/com/giraffe/match/navigation/MatchNavGraph.kt:
```diff
@@ -62,6 +64,9 @@ internal fun MatchNavGraph(
             navigateToYouTubePlayer = { videoId ->
                 navController.navigateToYouTubePlayer(videoId)
+            },
+            navigateToLoginScreen = {
```
consider reference function `navigateToLoginScreen = navController::navigateLoginScreen`

Mostafa-khaled21 committed:
Merge remote-tracking branch 'origin/develop' into feature/match-result

HendSayed25 commented on domain/media/src/main/java/com/giraffe/media/match/usecase/GetMatchingMoviesUseCase.kt:
```diff
@@ -0,0 +1,25 @@
+    private val repository: MatchRepository
+) {
+    suspend operator fun invoke(
+        genreIds: String,
```
Can you make data class contains all these parameter and only pass object from it ?

Mostafa-khaled21 replied:
Is it true in domain?

Mostafa-khaled21 replied:
I thought about it but I didn't do it so no one would think it was an entity

HendSayed25 commented on presentation/match/src/main/java/com/giraffe/match/screen/result/MatchResultViewModel.kt:
```diff
@@ -0,0 +1,377 @@
+    }
+
+    private fun checkLoginStatus() {
+        viewModelScope.launch {
```
Can you use safe execute rather than viewmodelScope ? 
To make the class have the same signature

dev-bilal-azzam commented on presentation/match/src/main/java/com/giraffe/match/screen/match_pager/MatchPagerViewModel.kt:
```diff
@@ -43,7 +64,9 @@ class MatchPagerViewModel @Inject constructor(
     fun onBackClicked() {
         viewModelScope.launch {
             if (_state.value.currentPage > 0) {
```
we are encabsulating the `private  _state ` with `public state` and `protected updateState(newState: S)`
to be able to read the state and to encabsulate updating state from the function `updateState` 
So why you are exposing this `_state `to the view model.

dev-bilal-azzam commented on presentation/match/src/main/java/com/giraffe/match/screen/result/MatchResultScreen.kt:
```diff
@@ -11,100 +12,112 @@ import androidx.compose.foundation.layout.statusBarsPadding
+    val state by viewModel.state.collectAsStateWithLifecycle()
+    val context = LocalContext.current
+
+    LaunchedEffect(viewModel.effect) {
```
consider using `EffectListener` utility for more consistency.

dev-bilal-azzam commented on presentation/match/src/main/java/com/giraffe/match/screen/result/MatchResultScreen.kt:
```diff
@@ -11,100 +12,112 @@ import androidx.compose.foundation.layout.statusBarsPadding
+            )
+
+            if (state.collectionBottomSheet != null) {
+                MovieCollectionsBottomSheet(
```
you have effects for all those navigation functions , so why you are sending them to the screen content again.
- those navigation function have to be called only when handling the effect. and the screen content must not call those navigation functions directly.
- But instead you have to trigger an interaction and then in the view model will send an effect responding to the interaction then the effect will be handled by calling the navigation function.

dev-bilal-azzam replied:
by the way all of this are already implemented but you just called the navigation function instead of calling interaction listener

dev-bilal-azzam commented on presentation/match/src/main/java/com/giraffe/match/screen/result/MatchResultScreen.kt:
Consider using the `BaseScreen`

Mostafa-khaled21 replied:
I already used it

Mostafa-khaled21 replied:
BaseScreenWithStates(
        isLoading = state.isLoading,
        isNoInternet = state.isNetworkError
    ) {

dev-bilal-azzam commented on presentation/match/src/main/java/com/giraffe/match/screen/result/MatchResultUiState.kt:
```diff
@@ -1,30 +1,92 @@
     val isGenericError: Boolean = false,
-    val isNetworkError: Boolean = false
-)
+    val isNetworkError: Boolean = false,
```
I think `isNoInternet` is more suitable name
- if you have `ShowError` Effect so why you add those 
```
    val isError: Boolean = false,
    val errorMsgRes: Int? = null,
    val isGenericError: Boolean = false
```
by the way you are not using them anywhere

dev-bilal-azzam commented on presentation/match/src/main/java/com/giraffe/match/screen/result/MatchResultViewModel.kt:
```diff
@@ -0,0 +1,377 @@
+) : BaseViewModel<MatchResultScreenState, MatchResultScreenEffect>(MatchResultScreenState()),
+    MatchResultInteractionListener {
+
+    private var _genreIds: List<Int> = emptyList()
```
why all those variables are not in the `MatchResultScreenState`

dev-bilal-azzam commented on presentation/match/src/main/java/com/giraffe/match/screen/result/MatchResultViewModel.kt:
```diff
@@ -0,0 +1,377 @@
+                    )
+                }
+            },
+            onError = { throwable, isNoInternet ->
```
I think this is better `onError = ::onError`

dev-bilal-azzam commented on presentation/match/src/main/java/com/giraffe/match/screen/result/MatchResultViewModel.kt:
```diff
@@ -0,0 +1,377 @@
+        )
+    }
+
+    private suspend fun mapMoviesToMatchResult(movies: List<Movie>): List<MatchResultModel> {
```
Since this function is calling a useCase, it is better to call it in safeExecute block

Mostafa-khaled21 replied:
try and catch because even when it was in safeExecute, an error would occur and it wouldn't retrieve any data if there was something null. I think this is the cause of the problem, as sometimes the details don't appear when we go in to see them in movie or series destails.

dev-bilal-azzam commented on presentation/match/src/main/java/com/giraffe/match/screen/result/MatchResultViewModel.kt:
```diff
@@ -0,0 +1,377 @@
+        }
+    }
+
+    private suspend fun mapSeriesToMatchResult(seriesList: List<Series>): List<MatchResultModel> {
```
the same as `mapMoviesToMatchResult`

Mostafa-khaled21 replied:
try and catch because even when it was in safeExecute, an error would occur and it wouldn't retrieve any data if there was something null. I think this is the cause of the problem, as sometimes the details don't appear when we go in to see them in movie or series destails.

dev-bilal-azzam commented on presentation/match/src/main/java/com/giraffe/match/composable/BaseScreenWithStates.kt:
```diff
@@ -0,0 +1,29 @@
+import com.giraffe.designsystem.composable.Progress
+
+@Composable
+fun BaseScreenWithStates(
```
consider adding `onRetryClick` to the `NoInternet` component

dev-bilal-azzam commented on presentation/match/src/main/java/com/giraffe/match/screen/result/MatchResultViewModel.kt:
```diff
@@ -0,0 +1,377 @@
+                    )
+                }
+            },
+            onError = { throwable, _ ->
```
move onSuccess and onError to separated function for more rdability, consistency

dev-bilal-azzam replied:
do this for safeExecute calls

dev-bilal-azzam commented on presentation/match/src/main/java/com/giraffe/match/screen/result/MatchResultViewModel.kt:
```diff
@@ -0,0 +1,377 @@
+    private fun executeIfLoggedIn(
+        block: () -> Unit,
+        ifNotLoggedIn: () -> Unit = { updateState { it.copy(isVisibleLoginBottomSheet = true) } }
+    ) {
```
I think using `safeExecute` here is better than `viewModelScope.launch` + `try/catch` block

Mostafa-khaled21 committed:
refactor match result view model and ui state

Mostafa-khaled21 committed:
Refactor MatchRemoteDataSource to use MatchApiService directly

Mostafa-khaled21 committed:
Refactor MatchResultScreen and related components

This commit refactors the `MatchResultScreen` and its related components:

- Updates `MatchNavGraph.kt` to use a method reference for `navigateToLoginScreen`.
- Adds an `onRetryClick` parameter to `BaseScreenWithStates.kt` to handle retry actions for no internet state.
- In `MatchResultScreen.kt`:
    - Replaces `LaunchedEffect` with a custom `EffectListener` for handling side effects.
    - Passes `onRetryClick` to `BaseScreenWithStates`.
    - Delegates navigation and interaction logic to the `viewModel`.

Mostafa-khaled21 committed:
Refactor MatchRepository variable name in use cases

Mostafa-khaled21 committed:
Enhance MatchResultInteractionListener and refactor MatchPagerViewModel and MatchRepositoryImpl

Mostafa-khaled21 committed:
refactor error handling in match result view model

Mostafa-khaled21 committed:
Merge remote-tracking branch 'origin/develop' into feature/match-result

Mostafa-khaled21 committed:
fix: handle null case in movie and series trailer url

Mostafa-khaled21 committed:
remove try catch from mapping functions in MatchResultViewModel

Mostafa-khaled21 committed:
fix match result screen scroll
